File: 0102binaryTreeLevelOrderTraversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function (root) {
  let ans = [];
  if (!root) return ans;

  function traverse(head, lvl) {
    if (!head) return;
    if (ans.length === lvl) ans.push([]);
    ans[lvl].push(head.val);
    traverse(head.left, lvl + 1);
    traverse(head.right, lvl + 1);
  }
  traverse(root, 0);
  return ans;
};

File: python/0102BinaryTreeLevelOrderTraversal.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:

    # BFS
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = []

        q = collections.deque()
        q.append(root)

        while q:
            qLen = len(q)
            level = []
            for i in range(qLen):
                node = q.popleft()
                if node:
                    level.append(node.val)
                    q.append(node.left)
                    q.append(node.right)

            if level:
                res.append(level)
        return res


# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.res = []

    # Recursive DFS
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        def trav(node, lvl):
            if not node:
                return

            # need to create new array
            if len(self.res) == lvl:
                self.res.append([node.val])
            else:
                self.res[lvl].append(node.val)
            trav(node.left, lvl + 1)
            trav(node.right, lvl + 1)

        trav(root, 0)
        return self.res
